### ISSUE TITLE
Excessive reconnects

### ISSUE THREAD

**Issue by i8beef** [Edubits/Zway-MQTT#23]
Hey man. I'm definitely still seeing excessive reconnects on my system, and I have a theory because I'm seeing a pattern. I suspect you have a race condition in your ping handling.

Your alive_timer and ping_timer have a relationship where ping_timer is 0.6 * alive_timer. This means, with the defaults of 30 second alive_timer timeout, you are sending a ping once every 18 seconds (ish, since it sets up the next timeout AFTER a successful response). In addition, when a ping is sent, you set a timeout to check for a successful ping response at "alive_timer" (30 seconds). However, on a successful ping response you don't CLEAR the original ping check, so it STILL fires after 30 seconds. As pings continue to run, and more of these checks are queued up to run every 30 seconds, eventually I think you can see an overlap where an old ping check runs in between a much later ping being sent and the corresponding ping response clearing the this._wait_for_pingresp. This triggers an (incorrect) Timeout, and then reconnect, and all those ping timeouts never get cleared, and continue running, and the problem just gets worse and worse as time goes on.

Recommendation: Instead of checking for a variable's value (this._wait_for_pingresp), set the variable to a timeout of the ping timeout check itself (this._wait_for_pingresp = setTimeout()). Then in the ping response, clearTimeout this var (clearTimeout(this._wait_for_pingresp)) instead of what you are doing. I believe this would be a nice concise way of stopping the proliferation of orphaned ping timeout checks causing incorrect timeouts. I'd probably actually rename that var if you did this to something more descriptive.

**Comment by i8beef** [comment]
Actually this is a good example of how to handle ping responses, right out of the zway codebases: https://github.com/Z-Wave-Me/O7/blob/master/O7-sock.js

If you are looking for an example, this is basically what I'm describing.

**Comment by i8beef** [comment]
So I took a stab at this in my repo... I'll watch it the next day or two to see if it stays stable. Hopefully you'll accept the pull request when I feel it's ready. I basically overhauled the entire MQTT library and changed the reconnect logic to be more reliable. There were a lot of scope weirdness issues with the MQTT library, and I was having such a hard time following it all that I just completely refactored it and walked everything to make sure the scopes were (more) correct. I may have missed one or two.

I also changed the reconnect logic to try and isolate the the reconnects and eliminate possibilities for multiple connections surviving and leaking resources. Added some more logging in a few areas as well so I could check on things through the flow. My branch has a hard coded var in it for turning on INFO logging which I'll turn off before PRing. 

I did NOT completely refactor the main module index.js. There are a lot of "var self = this" things littered everywhere in there that I'm pretty sure could be cleaned up with proper this scoping, unless I'm missing something, but they don't hurt anything right now.

Thus far it has already been MUCH better at reconnecting reliably and not leaking timers. I have tested the following and it seems much more reliable. I'm pretty sure this should make a lot of the stability issues I see people complain about (and I experience too) should be solved by my version:

1. All reconnect retries will slowly tend to a 60 second delay (just cleaned this up a bit)
2. Starting before MQTT is responding will trigger a series of reconnects to continue trying until it comes up, with careful attention paid to not spawning multiple retries at once, leaving timers running in the background, etc.
3. Failure at initial connect time no longer has a possibility to fall out of the reconnect loop and continue failing indefinitely until restart.
4. Failure after connect will trigger a log, correctly clean up all resources and timers, prevent multiple operations with timeouts from running at one time, and trigger the above reconnect logic reliably. 

I have NOT tested the following pieces:
1. Basic sub and pub seems to be working just fine, but QoS levels and message response handling I haven't really tested at all. All my stuff is QoS 0, but I noticed the MQTT library had facilities for dealing with 1 and 2 as well. I was careful, but could have fat fingered something there as I'm not exercising the code paths... I mean at least the PUB response ones work ok it seems though.

I changed a little bit of formatting for readability and added a lot of comments so I could keep track of what was happening as I refactored, but am happy to undo anything that doesn't fit your standard.

**Comment by Edubits** [comment]
Awesome @i8beef! Thanks a lot for your efforts. Recently I found out that the MQTT library indeed has overlapping requests and also sometimes tries to publish before it's connected, which is causing a similar issue. See also my last comment in #15. 

mqtt.js was an external library with some minimal changes to let it work in Zway, it really needed a cleanup and modernization, which I didn't have time for yet, so I'm really glad you took the effort! 

As I just moved to a new house, my own home automation is not yet functional again, so I'm unfortunately not able to test it. I hope to to have everything up and running next weekend. I'm looking forward to your PR! Could you do it against the 1.3 branch?

**Comment by i8beef** [comment]
I will make it against 1.3, for sure. Really the rewrite was more because the MQTT library was just a bit of a mess and I could not for the life of me track down where the connections were leaking, but  knew it has something to do with scoping. I just had to walk the entire library method by method to clean that all up, and I just reorganized things to make it easier as I went.

So far so good stability wise. I think the connection not REALLY opening, etc. stuff still happens sometimes with this, but it is so much more resilient and can accurately recover with reconnects from that state now that I'm now sure I'd do too much more to figure it out. 

If there are others having connection issues that are brave, I wouldn't mind some help testing this. Just need to download the index.js and lib/mqtt.js, replace them in the local copy under /opt/z-way-server/auotmation/userModules/MQTT and restart ZWay. Make backups first. It works great for me, but like I said, there could be others with setups slightly different from mine that I'm not setup to test.

If all is good still by next weekend, I will drop a PR on your plate. Cheers!

**Comment by i8beef** [comment]
Actually, it's been so stable thus far I just went ahead and made the PR to 1.3. I tested the reconnect logic for both MQTT coming up after ZWay and MQTT crashing or disconnecting in the middle a couple times and it handled both pretty well.

**Comment by Edubits** [comment]
I reviewed the changes this morning, seems good! (there are some differences in code style, but let's resolve that later) It also runs successful here locally. I will merge it into the 1.3 branch now!

**Comment by Edubits** [comment]
I found one issue: after you save configuration the module is restarted. This also triggers the reconnecting mechanism, which ends up in a loop. I resolved this with commit ae6ce05c9c03d6a74439ee7c1e9c56d726714944.

**Comment by i8beef** [comment]
I think that would be a good fix. I believe the example library I linked actually killed the onClose handler instead "to prevent recursive calls",, but I think I like your approach better, and Im not even sure their approach is addressing the same concern. Just be careful with the self = this stuff, as that was what I had to fix to get this stable. I suspect that the original author of that MQTT lib didn't have a decent grasp of closure usage here, but it didn't matter much with his usage because he wasn't reusing the object.
